File: trunk/main/src/com/seovic/identity/SequenceGenerator.java
/*
Copyright 2009 Aleksandar Seovic

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.seovic.identity;


import com.seovic.config.Configuration;

import com.tangosol.io.pof.PortableObject;
import com.tangosol.io.pof.PofReader;
import com.tangosol.io.pof.PofWriter;

import java.io.Serializable;
import java.io.IOException;

import java.lang.reflect.Constructor;

import java.util.concurrent.atomic.AtomicLong;


/**
 * An abstract base class for {@link IdentityGenerator} implementations
 * that generate sequential Long identifiers.
 *
 * @author Aleksandar Seovic  2009.05.27
 */
public abstract class SequenceGenerator
        implements IdentityGenerator<Long>, Serializable
    {
    // ---- constructors ----------------------------------------------------

    /**
     * Construct sequence generator.
     *
     * @param name       a sequence name
     * @param blockSize  the size of the sequence block to allocate at once
     */
    protected SequenceGenerator(String name, int blockSize)
        {
        this.m_name      = name;
        this.m_blockSize = blockSize;
        }

    public static SequenceGenerator create(String name)
        {
        return create(name, DEFAULT_BLOCK_SIZE);
        }

    public static SequenceGenerator create(String name, int blockSize)
        {
        try
            {
            Class type = Configuration.getSequenceGeneratorType();
            Constructor ctor = type.getConstructor(String.class, int.class);
            return (SequenceGenerator) ctor.newInstance(name, blockSize);
            }
        catch (Exception e)
            {
            throw new RuntimeException(e);
            }
        }

    
    // ---- properties ------------------------------------------------------

    public String getSequenceName()
        {
        return m_name;
        }

    public int getBlockSize()
        {
        return m_blockSize;
        }


    // ---- IdentityGenerator implementation --------------------------------

    /**
     * Return the next number in the sequence.
     *
     * @return the next number in the sequence
     */
    public synchronized Long generateIdentity()
        {
        SequenceBlock allocatedSequences = m_allocatedSequences;
        if (allocatedSequences == null || !allocatedSequences.hasNext())
            {
            m_allocatedSequences = allocatedSequences = allocateSequenceBlock();
            }
        return allocatedSequences.next();
        }


    // ---- abstract methods --------------------------------------------------

    /**
     * Allocate a new sequence block.
     *
     * @return block of sequential numbers
     */
    protected abstract SequenceBlock allocateSequenceBlock();


    // ---- data members ----------------------------------------------------

    /**
     * Default sequence block size.
     */
    public static final int DEFAULT_BLOCK_SIZE = 20;

    /**
     * Sequence name.
     */
    private final String m_name;

    /**
     * Sequence block size.
     */
    private final int m_blockSize;

    /**
     * Currently allocated block of sequences.
     */
    private SequenceBlock m_allocatedSequences;

    /**
 * Represents a block of sequential numbers.
     *
     * @author Aleksandar Seovic  2009.05.27
     */
    public static class SequenceBlock
            implements Serializable, PortableObject
        {
        // ---- constructors ----------------------------------------------------

        /**
         * Deserialization constructor (for internal use only).
         */
        public SequenceBlock()
            {
            }

        /**
         * Construct a new sequence block.
         *
         * @param first  first number in a sequence
         * @param last   last number in a sequence
         */
        public SequenceBlock(long first, long last)
            {
            m_next = new AtomicLong(first);
            m_last = last;
            }


        // ---- public methods --------------------------------------------------

        /**
         * Return <tt>true</tt> if there are avialable numbers in this
         * sequence block, <tt>false</tt> otherwise.
         *
         * @return <tt>true</tt> if there are avialable numbers in this
         *         sequence block, <tt>false</tt> otherwise
         */
        public boolean hasNext()
            {
            return m_next.longValue() <= m_last;
            }

        /**
         * Return the next available number in this sequence block.
         *
         * @return the next available number in this sequence block
         */
        public long next()
            {
            return m_next.getAndIncrement();
            }


        // ---- PortableObject implementation -----------------------------------

        /**
         * Deserialize object from the POF stream.
         *
         * @param reader  POF reader to use
         *
         * @throws IOException  if an error occurs
         */
        public void readExternal(PofReader reader)
                throws IOException
            {
            m_next = new AtomicLong(reader.readLong(0));
            m_last = reader.readLong(1);
            }

        /**
         * Serialize object into the POF stream.
         *
         * @param writer  POF writer to use
         *
         * @throws IOException  if an error occurs
         */
        public void writeExternal(PofWriter writer)
                throws IOException
            {
            writer.writeLong(0, m_next.longValue());
            writer.writeLong(1, m_last);
            }


        // ---- data members ----------------------------------------------------

        /**
         * The next assignable number within this sequence block.
         */
        private AtomicLong m_next;

        /**
         * The last assignable number within this sequence block.
         */
        private volatile long m_last;
        }
    }

File: main.net/src/Coherence.Tools/Identity/SequenceGenerator.cs
using System;
using System.Reflection;
using Seovic.Config;
using Tangosol.IO.Pof;

namespace Seovic.Identity
{
    /// <summary>
    /// An <see cref="IIdentityGenerator" /> implementation that generates 
    /// sequential long identifiers.
    /// </summary>
    /// <author>Aleksandar Seovic  2009.05.27</author>
    /// <author>Ivan Cikic  2009.09.29</author>
    public abstract class SequenceGenerator : IIdentityGenerator<long>
    {
        #region Constructors and factory methods

        /// <summary>
        /// Construct sequence generator.
        /// </summary>
        /// <param name="name">A sequence name.</param>
        /// <param name="blockSize">
        /// The size of the sequence block to allocate at once.
        /// </param>
        protected SequenceGenerator(string name, int blockSize)
        {
            m_name      = name;
            m_blockSize = blockSize;
        }

        /// <summary>
        /// Sequence generator factory method.
        /// </summary>
        /// <param name="name">A sequence name.</param>
        public static SequenceGenerator Create(string name)
        {
            return Create(name, DEFAULT_BLOCK_SIZE);
        }

        /// <summary>
        /// Sequence generator factory method.
        /// </summary>
        /// <param name="name">A sequence name.</param>
        /// <param name="blockSize">
        /// The size of the sequence block to allocate at once.
        /// </param>
        public static SequenceGenerator Create(string name, int blockSize)
        {
            Type type = Configuration.GetSequenceGeneratorType();
            ConstructorInfo ctor = type.GetConstructor(
                new Type[] {typeof(string), typeof(int)});
            return (SequenceGenerator) ctor.Invoke(new object[] {name, blockSize});
        }

        #endregion

        #region Properties

        /// <summary>
        /// Sequence name.
        /// </summary>
        public string SequenceName
        {
            get { return m_name; }
        }

        /// <summary>
        /// Sequence block size.
        /// </summary>
        public int BlockSize
        {
            get { return m_blockSize; }
        }

        #endregion

        #region IIdentityGenerator implementation

        /// <summary>
        /// Generates new identity instance.
        /// </summary>
        /// <returns>New identity instance</returns>
        public long GenerateIdentity()
        {
            lock (this)
            {
                SequenceBlock allocatedSequences = m_allocatedSequences;
                if (allocatedSequences == null || !allocatedSequences.HasNext())
                {
                    m_allocatedSequences = allocatedSequences = AllocateSequenceBlock();
                }
                return allocatedSequences.Next();
            }
        }

        #endregion

        #region Abstract methods

        /// <summary>
        /// Allocate a new sequence block.
        /// </summary>
        /// <returns>Block of sequential numbers</returns>
        protected abstract SequenceBlock AllocateSequenceBlock();

        #endregion

        #region Inner class: Sequence

        /// <summary>
        /// Named sequence.
        /// </summary>
        [Serializable]
        public class Sequence : IPortableObject
        {
            #region Constructors

            /// <summary>
            /// Deserialization constructor (for internal use only).
            /// </summary>
            public Sequence()
            {
            }

            /// <summary>
            /// Sequence constructor.
            /// </summary>
            /// <param name="name">Sequence name</param>
            public Sequence(string name)
            {
                m_name = name;
            }

            #endregion

            #region Public API

            /// <summary>
            /// Allocate a block of sequence numbers, starting from the last allocated
            /// sequence value.
            /// </summary>
            /// <param name="blockSize">The number of sequences to allocate</param>
            /// <returns>Allocated block of sequential numbers</returns>
            public SequenceBlock AllocateBlock(int blockSize)
            {
                long last = m_last;
                SequenceBlock block = new SequenceBlock(last + 1, last + blockSize);
                m_last = last + blockSize;

                return block;
            }

            #endregion

            #region Properties

            /// <summary>
            /// Return the sequence name.
            /// </summary>
            public string Name
            {
                get
                {
                    return m_name;
                }
            }

            /// <summary>
            /// Return the last allocated sequence number.
            /// </summary>
            public long Last
            {
                get
                {
                    return m_last;
                }
            }

            #endregion

            #region IPortableObject implementation

            public void ReadExternal(IPofReader reader)
            {
                m_name = reader.ReadString(0);
                m_last = reader.ReadInt64(1);
            }

            public void WriteExternal(IPofWriter writer)
            {
                writer.WriteString(0, m_name);
                writer.WriteInt64(1, m_last);
            }

            #endregion

            #region Data members

            /// <summary>
            /// Sequence name.
            /// </summary>
            private string m_name;

            /// <summary>
            /// The last allocated number from this sequence.
            /// </summary>
            private long m_last;

            #endregion
        }

        #endregion

        #region Inner class: SequenceBlock

        /// <summary>
        /// Represents a block of sequential numbers.
        /// </summary>
        [Serializable]
        public class SequenceBlock : IPortableObject
        {
            #region Constructors

            /// <summary>
            /// Deserialization constructor (for internal use only).
            /// </summary>
            public SequenceBlock()
            {
            }

            /// <summary>
            /// Construct a new sequence block.
            /// </summary>
            /// <param name="first">First number in a sequence</param>
            /// <param name="last">Last number in a sequence</param>
            public SequenceBlock(long first, long last)
            {
                m_next = first;
                m_last = last;
            }

            #endregion

            #region Public API

            /// <summary>
            /// Return <code>true</code> if there are avialable numbers in this
            /// sequence block, <code>false</code> otherwise.
            /// </summary>
            /// <returns>
            /// <code>true</code> if there are avialable numbers in this
            /// sequence block, <code>false</code> otherwise.
            /// </returns>
            public bool HasNext()
            {
                lock (this)
                {
                    return m_next <= m_last;
                }
            }

            /// <summary>
            /// Return the next available number in this sequence block.
            /// </summary>
            /// <returns>The next available number in this sequence block</returns>
            public long Next()
            {
                lock (this)
                {
                    return m_next++;
                }

            }

            #endregion

            #region IPortableObject implementation

            public void ReadExternal(IPofReader reader)
            {
                m_next = reader.ReadInt64(0);
                m_last = reader.ReadInt64(1);
            }

            public void WriteExternal(IPofWriter writer)
            {
                writer.WriteInt64(0, m_next);
                writer.WriteInt64(1, m_last);
            }

            #endregion

            #region Data members

            /// <summary>
            /// The next assignable number within this sequence block.
            /// </summary>
            private long m_next;

            /// <summary>
            /// The last assignable number within this sequence block.
            /// </summary>
            private long m_last;

            #endregion
        }

        #endregion

        #region Data members

        /// <summary>
        /// Default sequence block size.
        /// </summary>
        public static readonly int DEFAULT_BLOCK_SIZE = 20;

        /// <summary>
        /// Sequence name.
        /// </summary>
        private readonly String m_name;

        /// <summary>
        /// Sequence block size.
        /// </summary>
        private readonly int m_blockSize;

        /// <summary>
        /// Currently allocated block of sequences.
        /// </summary>
        private SequenceBlock m_allocatedSequences;

        #endregion
    }
}
